### DISCUSSION TITLE
Component on body tag?

### DISCUSSION THREAD

**Discussion by schmidt-oliver** [alpinejs/alpine#1274] [1. Help] [answered]
Hello community,

is it advisable to use an alpinejs component on body tag? Would it take a huge effect on performence, especially on a page with a huge DOM?

**Goal:**
If a page overlay is displayed, scrolling should be disabled. Technically: add and remove a css class at the body tag at a given event.

Example **with** component on body tag: https://jsfiddle.net/b59nm0ay/1/
Example **without** component on body tag: https://jsfiddle.net/wbrma6p8/

Personally I would prefer the first way with a component on body tag, cause of writing less and more consistent code.

Any experiences or recommendations? Thank you!

**Comment by ryangjchandler** [comment, marked as answer]
@schmidt-oliver If the DOM is huge, it _could_ have a performance impact because Alpine will walk each node in the subtree (child elements, etc). You should have a small component inside of the body with no children that listens for the event and then just uses `document.body` to mutate the DOM.

**Reply by schmidt-oliver** [reply]
But won't Alpine still walk through the entire DOM to find components and directives? Does it walk through the subtree of each component individually?

**Reply by ryangjchandler** [reply]
> But won't Alpine still walk through the entire DOM to find components and directives? Does it walk through the subtree of each component individually?

Alpine will only walk through the DOM tree for elements with an `x-data` attribute. If you have something like this:

```html
<body>
    <div x-data></div>

    <div></div>
</body>
```

Alpine won't walk the second `<div>`, only the first one.

**Reply by schmidt-oliver** [reply]
Thank you. I like your idea with a component without any children listining for the event.

Found this lesson https://laracasts.com/series/building-alpinejs very helpfull to get a bit clearer how Alpine works under the hood.

**Comment by KevinBatdorf** [comment]
There's also a performance issue from Alpine running a callback on every mutation. On a site with no interactions this is probably okay, but if things are changing often, you could see slow downs.

**Comment by earthboundkid** [comment]
Neat way to do this [with Alpine 3](https://jsfiddle.net/px3tvwf4/1/):

```html
    <data 
      x-data="{ visible: false }" 
      @toggle.window="visible = ! visible" 
      x-effect="document.body.classList.toggle('overlayed', visible)"></data>
    <div class="content">
      <button 
        type="button" 
        class="" 
        x-data 
        @click.stop.prevent="$dispatch('toggle')"
       >
        Toggle Overlay
      </button>
    </div>
```

**Comment by danddanddand** [comment]
I would strongly recommend using `body-scroll-lock` package.

I originally tried something similar to your solution but there were edge cases that were hard to handle (can't remember but maybe with scroll bar gap or Safari or iOS?). Plus the code is quite a bit cleaner because your modal or mobile menu open() simply calls

```
disableBodyScroll( $refs.menuContent, {
    reserveScrollBarGap: true,
} );
```

and your close() would enable it again. 

`enableBodyScroll( $refs.menuContent );`

Here is our full example (sorry it's ts). Big advantage is no css/events/component on body and it works great. 
```
import { disableBodyScroll, enableBodyScroll } from "body-scroll-lock";

export interface MobileMenuComponentProps {
    isOpen: boolean;
    init: () => void;
    open: () => void;
    close: () => void;
    toggle: () => void;
}

export type MobileMenuComponent = () => MobileMenuComponentProps;

export const mobileMenuComponent: AlpineComponent<MobileMenuComponentProps> = () => {
    return {
        isOpen: false,
        init: function () {
            this.$refs!.menuContent.style.height = `${window.innerHeight }px`;
        },
        open: function () {
            this.isOpen = true;
            disableBodyScroll( this.$refs!.menuContent, {
                reserveScrollBarGap: true,
            } );
        },
        close: function () {
            this.isOpen = false;
            enableBodyScroll( this.$refs!.menuContent );
        },
        toggle: function () {
            if ( this.isOpen ) {
                this.close();
            } else {
                this.open();
            }
        },
    };
};

```
